File: packages/adapters/src/email.ts
/**
 * Email contract — transactional email: send, template, domain verification.
 *
 * Mirrors `fiducial_adapters::email` in Rust. Designed against Resend, SES,
 * and Cloudflare Email Routing.
 *
 * Unlike `D1Database`/`R2Storage`, `ResendEmail` is **not** binding-reached —
 * it is a secret key (`env.RESEND_API_KEY`, set via `wrangler secret put`)
 * and a POST request over plain HTTPS, so it works from any TypeScript runtime
 * with `fetch`. Same boundary as `Turnstile`.
 */

/** An outbound email message. */
export interface Message {
  /** `"Name <addr>"` or bare `"addr"`. */
  from: string;
  /** One or more recipient addresses. */
  to: string[];
  subject: string;
  /** HTML body (required). */
  html: string;
  /** Optional plain-text alternative. */
  text?: string;
  /** Optional reply-to address. */
  replyTo?: string;
}

export class EmailError extends Error {
  constructor(
    message: string,
    public readonly code?: string,
  ) {
    super(message);
    this.name = "EmailError";
  }
}

/** Transactional email contract. */
export interface Email {
  /**
   * Send a single transactional message.
   *
   * Returns the provider-assigned message ID on success.
   */
  send(message: Message): Promise<string>;
}

// ── None implementation ───────────────────────────────────────────────────────

/**
 * No-op email sender — all sends succeed silently, returning an empty ID.
 *
 * Switching to Resend in production is a config change, not a refactor.
 */
export class NoneEmail implements Email {
  // Accepts and ignores `env` so every vendor class in this contract shares
  // one constructor shape for the generated factory to call uniformly.
  constructor(_env?: unknown) {}

  async send(_message: Message): Promise<string> {
    return "";
  }
}

// ── Resend ────────────────────────────────────────────────────────────────────

const RESEND_SEND_URL = "https://api.resend.com/emails";

interface ResendSendResponse {
  id: string;
}

/**
 * Resend — sends transactional email via `POST https://api.resend.com/emails`.
 *
 * Secret-reached: `env.RESEND_API_KEY`, set with `wrangler secret put`.
 * Constructor throws `EmailError` when the key is absent, with the
 * `wrangler secret put` command in the message — a misconfigured product fails
 * at construction rather than at the first send.
 *
 * `fetchImpl` defaults to global `fetch` and exists so tests can pass a stub
 * instead of monkeypatching `globalThis.fetch`.
 *
 * Resend's documented default rate limit is 10 req/s per team. HTTP 429
 * is mapped onto `EmailError` because the Rust contract already carries a
 * `RateLimited` variant and this is a real path, not a defensive one.
 */
export class ResendEmail implements Email {
  private readonly apiKey: string;
  private readonly fetchImpl: typeof fetch;

  constructor(
    env: { RESEND_API_KEY?: string },
    fetchImpl: typeof fetch = fetch,
  ) {
    if (!env?.RESEND_API_KEY) {
      throw new EmailError(
        "ResendEmail: no `RESEND_API_KEY` on env — set it with " +
          "`wrangler secret put RESEND_API_KEY`",
      );
    }
    this.apiKey = env.RESEND_API_KEY;
    this.fetchImpl = fetchImpl;
  }

  async send(message: Message): Promise<string> {
    let res: Response;
    try {
      res = await this.fetchImpl(RESEND_SEND_URL, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          from: message.from,
          to: message.to,
          subject: message.subject,
          html: message.html,
          text: message.text,
          reply_to: message.replyTo,
        }),
      });
    } catch (err) {
      throw new EmailError(`ResendEmail request failed: ${String(err)}`);
    }

    if (res.status === 429) {
      throw new EmailError("ResendEmail: rate limited (HTTP 429)", "rate_limited");
    }

    if (!res.ok) {
      throw new EmailError(
        `ResendEmail returned HTTP ${res.status}`,
        String(res.status),
      );
    }

    const data = (await res.json()) as ResendSendResponse;
    return data.id;
  }
}

File: crates/fiducial-adapters/src/email.rs
//! Email contract — transactional email: send, template, domain verification.
//!
//! Designed against the narrowest interface shared by Resend, SES and
//! Cloudflare Email Routing. Vendor-specific features (Resend batch sends,
//! SES configuration sets, Cloudflare's SPF/DKIM helpers) stay on the vendor.

use thiserror::Error;

use crate::BoxFuture;

/// An outbound email message.
#[derive(Debug, Clone)]
pub struct Message {
    /// `name <addr>` or bare `addr`.
    pub from: String,
    /// One or more recipient addresses.
    pub to: Vec<String>,
    pub subject: String,
    /// HTML body (required). Plain-text alt is derived by stripping tags if
    /// the vendor supports it; provide `text` explicitly to control it.
    pub html: String,
    /// Optional plain-text alternative.
    pub text: Option<String>,
    /// Optional reply-to address.
    pub reply_to: Option<String>,
}

/// Errors an [`Email`] implementation may return.
#[derive(Debug, Error)]
pub enum EmailError {
    #[error("send rejected by provider: {reason}")]
    Rejected { reason: String },
    #[error("invalid address: {addr}")]
    InvalidAddress { addr: String },
    #[error("rate limited — retry after {retry_after_secs}s")]
    RateLimited { retry_after_secs: u64 },
    #[error("{0}")]
    Other(String),
}

/// Transactional email contract.
pub trait Email: Send + Sync {
    /// Send a single transactional message.
    ///
    /// Returns the provider-assigned message ID on success.
    fn send<'a>(&'a self, message: &'a Message) -> BoxFuture<'a, Result<String, EmailError>>;
}

// ── None implementation ───────────────────────────────────────────────────────

/// No-op email sender — all sends succeed silently, returning an empty ID.
///
/// Useful in development and testing: the product is wired for email from the
/// first line, and switching to Resend in production is a config change, not a
/// refactor.
pub struct NoneEmail;

impl Email for NoneEmail {
    fn send<'a>(&'a self, _message: &'a Message) -> BoxFuture<'a, Result<String, EmailError>> {
        Box::pin(std::future::ready(Ok(String::new())))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn email() -> Box<dyn Email> {
        Box::new(NoneEmail)
    }

    fn msg() -> Message {
        Message {
            from: "hello@example.com".into(),
            to: vec!["user@example.com".into()],
            subject: "Test".into(),
            html: "<p>Hi</p>".into(),
            text: None,
            reply_to: None,
        }
    }

    #[tokio::test]
    async fn send_succeeds_silently() {
        let id = email().send(&msg()).await.unwrap();
        assert_eq!(id, "");
    }

    #[test]
    fn none_email_is_object_safe() {
        let _: Box<dyn Email> = Box::new(NoneEmail);
    }
}
